### PR TITLE
Serializer lights punctual

### PR DESCRIPTION
**Pull request by kcoley**
### glTF Serializer
- Added support for exporting `KHR_lights_punctual`
- change `shouldExportTransformNode` callback to `shouldExportNode`

### CODE REVIEW

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                let babylonLight: ShadowLight = babylonNode;
+                let light: ILight;
+
+                let lightType = (
+                    babylonLight.getTypeID() == Light.LIGHTTYPEID_POINTLIGHT ? LightType.POINT : (
+                        babylonLight.getTypeID() == Light.LIGHTTYPEID_DIRECTIONALLIGHT ? LightType.DIRECTIONAL : (
+                            babylonLight.getTypeID() == Light.LIGHTTYPEID_SPOTLIGHT ? LightType.SPOT : null
+                        )));
+                if (lightType == null) {
+                    Tools.Warn(`${context}: Light ${babylonLight.name} is not supported in {NAME}`);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]:
```suggestion
                    Tools.Warn(`${context}: Light ${babylonLight.name} is not supported in ${NAME}`);
```

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                        light.range = babylonLight.range;
+                    }
+
+                    if (lightType === LightType.SPOT) {
+                        let babylonSpotLight = babylonLight as SpotLight;
+                        if (!(babylonSpotLight.angle == Math.PI / 2.0)) {
+                            if (light.spot == null) {
+                                light.spot = {};
+                            }
+                            light.spot.outerConeAngle = babylonSpotLight.angle / 4;
```
Comment (bghgary) [inline serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: This doesn't seem right. It should be the opposite of the loader, no?

```suggestion
                            light.spot.outerConeAngle = babylonSpotLight.angle / 2;
```

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                        }
+                        if (!(babylonSpotLight.innerAngle == 0)) {
+                            if (light.spot == null) {
+                                light.spot = {};
+                            }
+                            light.spot.innerConeAngle = babylonSpotLight.innerAngle / 2;
+                        }
+                        let rotation = babylonSpotLight.getRotation().clone();
```
Comment (bghgary) [inline serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: Can be shared with the code below? `getRotation` is on `ShadowLight`

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                            node.rotation = Quaternion.FromEulerAngles(rotation.x, rotation.y, rotation.z).asArray();
+                        }
+                    }
+                    if (this._lights == null) {
+                        this._lights = {
+                            lights: []
+                        };
+                    }
+
+                    this._lights.lights.push(light);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: can lights be shared or instanced?

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 2 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: @bghgary yeah should be / 2

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 3 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: @bghgary I can bump it up. I wasn't sure if it made sense to have it on PointLight, but it should just be ignored by engines and can save code duplication.

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 4 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: @bghgary good point.  I can check if the light has already been serialized first.

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        outerConeAngle?: number;
+    };
+}
+
+interface ILights {
+    lights: ILight[];
+}
+
+/**
+ * [Specification](https://github.com/KhronosGroup/glTF/blob/1048d162a44dbcb05aefc1874bfd423cf60135a6/extensions/2.0/Khronos/KHR_lights_punctual/README.md) (Experimental)
```
[the remark this answers is not in the document]
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: @bghgary sure I'll update the README link

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 1 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: @bghgary thanks, missed that one

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 3 above)
Comment (bghgary) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: You can always put another if check around it.

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 4 above)
Comment (bghgary) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: Ideally, I think we should check if the lights have the same properties and don't export a light with the same properties.

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 3 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: @bghgary actually, I was handling rotation incorrectly.  I will make an update to use the direction on the spotlight and directional light instead.

File: dist/preview release/what's new.md [hunk 6]
```diff
@@ -112,6 +112,9 @@
 
 ### glTF Serializer
 
+- Added support for exporting `KHR_lights_punctual`
+- change `shouldExportTransformNode` callback to `shouldExportNode`
```
[the remark this answers is not in the document]
Comment (kcoley) [reply dist/preview release/what's new.md]: I moved this to the breaking change section

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 3 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: update has been made

File: serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts (hunk 4 above)
Comment (kcoley) [reply serializers/src/glTF/2.0/Extensions/KHR_lights_punctual.ts]: This looks to be a bit complex to implement, but maybe an optimization for a future update.

**Comment by kcoley** [comment]
@bghgary I added some unit/integration tests for the lights

File: dist/preview release/what's new.md [hunk 7]
```diff
@@ -223,3 +225,4 @@
 - No more `babylon.worker.js` javascript following the lack of usage from the feature ([Sebavan]
 (https://github.com/Sebavan))
 - No more `Primitive Geometries` as they were not in use since 2.0 ([Sebavan](https://github.com/Sebavan))
+- change `shouldExportTransformNode` callback in glTF serializer options to `shouldExportNode`([kcoley](https://github.com/kcoley))
```
Comment (bghgary) [inline dist/preview release/what's new.md]:
```suggestion
- Change `shouldExportTransformNode` callback in glTF serializer options to `shouldExportNode`([kcoley](https://github.com/kcoley))
```

**Comment by kcoley** [comment]
@bghgary looks like the spotlight rotation serialization is not correct when serializing to glTF.  Looking into it.

**Comment by Drigax** [comment]
Hey gonchar, I just noticed your comment, this pull request has been closed for a while so we missed your comment. If you're still running into your issue, can you open a new issue with more detail?

**Pull request merged**